task.c

```diff
@@ -0,0 +1,120 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "task.h"
+#include "common.h"
+#include "sockets.h"
+
+/* Tasks are stored in a consecutive chunk of memory, the first
+ * sizeof(task_spec) bytes are arranged acording to the struct
+ * task_spec. Then there is an array of task_args of length
+ * (num_args + num_returns), and then follows the data of
+ * pass-by-value arguments of size args_value_size. The offsets in the
+ * task_arg.val are with respect to the end of the struct, i.e.
+ * with respect to the address &task_spec.ids[0]. */
+
+typedef struct {
+  /* Either ARG_BY_REF or ARG_BY_VAL. */
+  int8_t type;
+  union {
+    unique_id obj_id;
+    struct {
+      /* Offset where the data associated to this arg is located relative
+       * to &task_spec.ids[0]. */
+      ptrdiff_t offset;
+      int64_t length;
+    } val;
+  };
+} task_arg;
+
+struct task_spec_impl {
+  function_id func_id;
+  /* Total number of arguments. */
+  int64_t num_args;
+  /* Position of the argument that is currently constructed. */
+  int64_t arg_index;
+  /* Number of return values. */
+  int64_t num_returns;
+  /* Number of bytes the pass-by-value arguments are occupying. */
+  int64_t args_value_size;
+  /* Offset in the argument data. */
+  int64_t args_value_offset;
+  /* Argument and return IDs. */
+  task_arg ids[0];
```

Comment:
These aren't all ids, right? Maybe call them "args_and_returns"? And for the comment "Arguments and return IDs and values"